components/constellation/constellation.rs

```diff
@@ -1922,49 +1926,64 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
         }
     }
 
-    fn traverse_frame_to_pipeline(&mut self, frame_id: FrameId, next_pipeline_id: PipelineId) {
+    fn traverse_frame_to_timestamp(&mut self, frame_id: FrameId, instant: Instant) {
         // Check if the currently focused pipeline is the pipeline being replaced
         // (or a child of it). This has to be done here, before the current
         // frame tree is modified below.
         let update_focus_pipeline = self.focused_pipeline_in_tree(frame_id);
 
-        let prev_pipeline_id = match self.frames.get_mut(&frame_id) {
+        let (prev_pipeline_id, next_pipeline_id, entry) = match self.frames.get_mut(&frame_id) {
             Some(frame) => {
-                let prev = frame.current.pipeline_id;
-
-                // Check that this frame contains the pipeline passed in, so that this does not
-                // change Frame's state before realizing `next_pipeline_id` is invalid.
-                if frame.next.iter().find(|entry| next_pipeline_id == entry.pipeline_id).is_some() {
-                    frame.prev.push(frame.current.clone());
-                    while let Some(entry) = frame.next.pop() {
-                        if entry.pipeline_id == next_pipeline_id {
-                            frame.current = entry;
-                            break;
-                        } else {
-                            frame.prev.push(entry);
-                        }
+                let prev_pipeline_id = frame.pipeline_id;
+                let mut entry = frame.current();
+
+                if instant > frame.instant {
```

Comment:
This is much cleaner! Hopefully there can never be a case where an `Instant` is passed that is not in the `Frame` history. I can't think of any cases where that could happen, but it would be pretty destructive if it did happen.